### PR TITLE
11 Networking.Persistence Module with DAO abstract class, implementations and builder

### PR DESCRIPTION
**Pull request by SverreNystad**
# The network.persistence module

**This PR adds the following to the project:**
* **DAO** Abstract base class
* **DAOBuilder** for creation of DAOs
* **FirebaseDAO** for multiplayer (NOT YET IMPLEMENTED)
* **LocalDAO** for singleplay (NOT YET IMPLEMENTED)
* **MockDAO** used for testing

## There are also tests for this functionality
To run all the tests:
```cmd
gradlew test
```

### CODE REVIEW

**Comment by SverreNystad** [comment]
Found a tutorial on how to add firebase. So I can probably set it later today

File: core/src/com/softwarearchitecture/networking/persistence/DAO.java [hunk 1]
```diff
@@ -1,5 +1,102 @@
...
+ * This abstract base class abstracts the interaction with the data storage system, allowing for operations
+ * on stored data without exposing the details of the implementation. It provides a standardized
+ * method to access, modify, and manage data across different types of storage mechanisms.
+ * 
+ * @param <T> The type of the object managed by this DAO.
+ * @param <K> The type of the primary key used to identify objects of type T.
+ * 
+ * @see <a href="https://en.wikipedia.org/wiki/Data_access_object">Data Access Object on Wikipedia</a>
+ */
+public abstract class DAO<T, K> {
```
Comment (mvbryne) [inline core/src/com/softwarearchitecture/networking/persistence/DAO.java:16]: Is this the most intuitive way to implement the type before the key? Maybe DAO<K, T>

File: core/src/com/softwarearchitecture/networking/persistence/LocalDAO.java
Comment (mvbryne) [inline core/src/com/softwarearchitecture/networking/persistence/LocalDAO.java:1]: persistence not persistance

File: core/src/com/softwarearchitecture/networking/persistence/DAO.java (hunk 1 above)
Comment (SverreNystad) [reply core/src/com/softwarearchitecture/networking/persistence/DAO.java:16]: Yes like with how Hashmaps does it. I totally agree

File: core/src/com/softwarearchitecture/networking/persistence/MockDAO.java
[the remark this answers is not in the document]
Comment (mattiastofte) [reply core/src/com/softwarearchitecture/networking/persistence/MockDAO.java:1]: Or i don't see the reason to test something that won't be used, if that makes sense

**Review by Artewald — APPROVED** [review]
Looks good :D

**Pull request merged**